Better selecting!

JorisGoosen opened the pull request:
started work on jasp-stats/INTERNAL-jasp#2495

this PR will:
- allow for selecting disjunct sections of the data
- actions on selected stuff should only do things for those selected
- copy pasting within jasp will only copy those cells selected
- to spreadsheet it will copy nonselected cells as ""
- cmd/shift clicking now should work

I couldnt get the nice shift-select on hover to work properly for now but that is way less important than being able to change columntypes or reversing levels so here we are.
Ive turned on a build of `betterSelecting` for @JohnnyDoorn to review, if the static server with the nightlies does not go down in the meantime

JorisGoosen committed:
started work on https://github.com/jasp-stats/INTERNAL-jasp/issues/2495

improve selection and Buttons with icons resolution

JorisGoosen committed:
slightly better behaved

JorisGoosen committed:
selecting works reasonable

JorisGoosen committed:
make sure toggling whole columns/rows properly

next time continue with undostack!

JorisGoosen committed:
massive speedup to copy pasting lots of data

JorisGoosen committed:
some improvements to copy-paste

JorisGoosen committed:
break build as a reminder to use only selected data when pasting within jasp

JorisGoosen committed:
overwrite only those cells that were copied

also make sure that changing columntype or reversing value etc is only done on selected columns, not on everything between min and max

JorisGoosen committed:
make sure you can remove disjunct sets of rows or columns

JorisGoosen committed:
remove useless copy of UndoStack pointer

JorisGoosen committed:
dont crash on undoing big paste

JorisGoosen committed:
make sure to not have the edithighlight linger

boutinb commented:
Several issues found:

- An extra insert negative rows in the Undo stack appears when copy paste some values. I solved it in ExpandDataProxyModel::_expandIfNecessary (https://github.com/jasp-stats/jasp-desktop/pull/5466/commits/f5d507065e21953c21b99848a411038bf7ca7efc).
- The VariableWindow does not show the selected variable in the Data view. It shows the right variable only if the header of the column is clicked, not if a cell is clicked.
<img width="535" alt="Screenshot 2024-04-19 at 15 34 53" src="https://github.com/jasp-stats/jasp-desktop/assets/503332/d4346b48-823d-42df-8d6c-f022db72f81a">

- If you set a value on an extra row, this value disappears.
- If you copy-paste some values on a column, and some of these values add some labels, when undoing the values are reset to their original values, but the labels stay even if no value uses them. This is not the case when just setting a new value (by typing directly in the cell), and undoing it.

JorisGoosen committed:
fix supporters spacing a bit
also fix selecting of column in variableswindow on cellclick

JorisGoosen commented:
Also a point I found: doing selext all on certain data (probably just big, as I used the data from https://github.com/jasp-stats/INTERNAL-jasp/issues/2498) is crazy slow.

JorisGoosen committed:
several fixes

- dont insert rows if you dont need to
- check for positive count is not a bad idea indeed
- use old PasteSpreadSheetCommand code because the "updated version" was way too buggy

JorisGoosen committed:
speed up selectin massively by not uselessly checking every cell to get min/max range...

JorisGoosen committed:
sort timers by user duration

JorisGoosen committed:
fix progressbar for readstat

JorisGoosen committed:
only insert rows if necessary

JorisGoosen committed:
more fixes

JorisGoosen committed:
copy only what is needed

JorisGoosen committed:
better cutting and pasting

JorisGoosen committed:
ahum

JorisGoosen committed:
combine cut all into one undo action

fix cutting disjunct selection

JorisGoosen committed:
better cut all

JorisGoosen commented:
Alright I guess this is ready for review again, ive turned on a build for macos arm  for Johnny in case he wants to have a look

JorisGoosen commented:
I will work on a speedup for big copies in a separate branch though

JorisGoosen committed:
make sure that both setting a value in a virtual cell expands properly, and that the same goes for pasting spreadsheet

JorisGoosen committed:
multi-column operations should properly handle being called on something not in the selection

JorisGoosen committed:
fix weird missing value problem
